### PR TITLE
Remove DoFHandlerType from MatrixFree::reinit()

### PR DESCRIPTION
**Pull request by peterrum**
References https://github.com/dealii/dealii/pull/10334#issuecomment-644603560 and #10333.

### CODE REVIEW

File: include/deal.II/matrix_free/matrix_free.h [hunk 1]
```diff
@@ -612,10 +612,21 @@ class MatrixFree : public Subscriptor
...
   reinit(const Mapping<dim> &                                   mapping,
-         const std::vector<const DoFHandlerType *> &            dof_handler,
+         const std::vector<const DoFHandler<dim> *> &           dof_handler,
+         const std::vector<const AffineConstraints<number2> *> &constraint,
+         const std::vector<QuadratureType> &                    quad,
+         const AdditionalData &additional_data = AdditionalData());
+
+  /**
+   * Initializes the data structures. Same as above, but  using hp::DoFHandlers.
+   */
```
Comment (masterleinad) [inline include/deal.II/matrix_free/matrix_free.h:625]:
```suggestion
     @deprecated Use the overload taking a DoFHandler object instead.
   */
```

File: include/deal.II/matrix_free/matrix_free.h [hunk 2]
```diff
@@ -624,9 +635,19 @@ class MatrixFree : public Subscriptor
...
+  template <typename QuadratureType, typename number2>
+  void
+  reinit(const std::vector<const DoFHandler<dim> *> &           dof_handler,
+         const std::vector<const AffineConstraints<number2> *> &constraint,
+         const std::vector<QuadratureType> &                    quad,
+         const AdditionalData &additional_data = AdditionalData());
+
+  /**
+   * Initializes the data structures. Same as above, but  using hp::DoFHandlers.
+   */
```
Comment (masterleinad) [inline include/deal.II/matrix_free/matrix_free.h:647]:
```suggestion
     @deprecated Use the overload taking a DoFHandler object instead.
   */
```

File: include/deal.II/matrix_free/matrix_free.h [hunk 3]
```diff
@@ -638,10 +659,22 @@ class MatrixFree : public Subscriptor
...
-         const std::vector<const DoFHandlerType *> &            dof_handler,
+         const std::vector<const DoFHandler<dim> *> &           dof_handler,
+         const std::vector<const AffineConstraints<number2> *> &constraint,
+         const std::vector<IndexSet> &      locally_owned_set,
+         const std::vector<QuadratureType> &quad,
+         const AdditionalData &             additional_data = AdditionalData());
+
+  /**
+   * Initializes the data structures. Same as above, but  using hp::DoFHandlers.
+   */
```
Comment (masterleinad) [inline include/deal.II/matrix_free/matrix_free.h:673]:
```suggestion
     @deprecated Use the overload taking a DoFHandler object instead.
   */
```

File: include/deal.II/matrix_free/matrix_free.h [hunk 4]
```diff
@@ -654,10 +687,21 @@ class MatrixFree : public Subscriptor
...
   reinit(const Mapping<dim> &                                   mapping,
-         const std::vector<const DoFHandlerType *> &            dof_handler,
+         const std::vector<const DoFHandler<dim> *> &           dof_handler,
+         const std::vector<const AffineConstraints<number2> *> &constraint,
+         const QuadratureType &                                 quad,
+         const AdditionalData &additional_data = AdditionalData());
+
+  /**
+   * Initializes the data structures. Same as above, but  using hp::DoFHandlers.
+   */
```
Comment (masterleinad) [inline include/deal.II/matrix_free/matrix_free.h:700]:
```suggestion
     @deprecated Use the overload taking a DoFHandler object instead.
   */
```

File: include/deal.II/matrix_free/matrix_free.h [hunk 5]
```diff
@@ -666,9 +710,19 @@ class MatrixFree : public Subscriptor
...
   void
-  reinit(const std::vector<const DoFHandlerType *> &            dof_handler,
+  reinit(const std::vector<const DoFHandler<dim> *> &           dof_handler,
+         const std::vector<const AffineConstraints<number2> *> &constraint,
+         const QuadratureType &                                 quad,
+         const AdditionalData &additional_data = AdditionalData());
+
+  /**
+   * Initializes the data structures. Same as above, but  using hp::DoFHandlers.
+   */
```
Comment (masterleinad) [inline include/deal.II/matrix_free/matrix_free.h:722]:
```suggestion
     @deprecated Use the overload taking a DoFHandler object instead.
   */
```

File: include/deal.II/matrix_free/matrix_free.h (hunk 1 above)
Comment (peterrum) [reply include/deal.II/matrix_free/matrix_free.h:627]: You are right! I have added the comments!

**Comment by peterrum** [comment]
The two places with `DoFHandlerType` in `matrix_free.h` remaining:

https://github.com/dealii/dealii/blob/57525d52611bae4e1f1517c6f32201722229995e/include/deal.II/matrix_free/matrix_free.h#L1708-L1718

https://github.com/dealii/dealii/blob/57525d52611bae4e1f1517c6f32201722229995e/include/deal.II/matrix_free/matrix_free.h#L2822-L2858

My guess is that the second one can go. What about the first one?

**Comment by kronbichler** [comment]
They can both go away; the first is however an incompatibility in terms of the request by the user, so the question is whether we simply say we ignore that or do deprecation and then have another interface instead. I think the feature is rather new and the non-default dof handler is not common, so it would probably not hurt too much. On the other hand, does it hurt to keep it? Not very much I would say.

**Comment by peterrum** [comment]
> On the other hand, does it hurt to keep it? Not very much I would say.

I agree. But, nevertheless, the goal till the next release should be to try to eliminate all possible `DoFHandlerTypes` as long as they not cause any major incompatibility (see also #10333).

**Pull request merged**